Fix signed division bias for power-of-two divisor in BBQ JIT

ddegazio opened the pull request:
#### acb670a0a6d0437d39ba904083ee21ad89799a43
<pre>
Fix signed division bias for power-of-two divisor in BBQ JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=256569">https://bugs.webkit.org/show_bug.cgi?id=256569</a>
rdar://108073980

Reviewed by Yusuke Suzuki.

Fixes instruction selection for integer division and remainder
in BBQ JIT when the divisor is a constant power of two, with new
tests for the edge cases previously not covered by our suite.

* JSTests/wasm/stress/divide-by-constant-power-of-two.js: Added.
(async test):
* JSTests/wasm/stress/remainder-by-constant-power-of-two.js: Added.
(async test):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::add64):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitModOrDiv):

Canonical link: <a href="https://commits.webkit.org/263925@main">https://commits.webkit.org/263925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1a0b38a5200545f4081aa8346653b5bf62b800b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6109 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7672 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6457 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6247 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9331 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6219 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7734 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3712 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5506 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13415 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5133 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5577 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5584 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7818 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5701 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6055 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4957 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6230 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5476 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1488 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1450 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9616 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6403 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5844 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1613 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -300,6 +300,23 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
         m_assembler.add<64>(dest, src, dataTempRegister);
     }
 
+    void add64(TrustedImm64 imm, RegisterID src, RegisterID dest)
```
Don't you also need to add this emitter for x64?  I don't see it in the codebase yet.

MenloDorian replied:
Nevermind.  It looks like this is only used in the `!CPU(X86_64)` section of the code.

ddegazio committed:
Fix signed division bias for power-of-two divisor in BBQ JIT
https://bugs.webkit.org/show_bug.cgi?id=256569
rdar://108073980

Reviewed by Yusuke Suzuki.

Fixes instruction selection for integer division and remainder
in BBQ JIT when the divisor is a constant power of two, with new
tests for the edge cases previously not covered by our suite.

* JSTests/wasm/stress/divide-by-constant-power-of-two.js: Added.
(async test):
* JSTests/wasm/stress/remainder-by-constant-power-of-two.js: Added.
(async test):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::add64):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitModOrDiv):

Canonical link: https://commits.webkit.org/263925@main
